### PR TITLE
runk: add tests for runk

### LINKED ISSUE #4882: Add tests cases for runk
Add tests for runk as an OCI runtime.

### PR DESCRIPTION
**Pull request by cyyzero**
Use runk as the default OCI runtime of containerd to validate runk's
commands. Tests includes creating containers, exec a process, kill and
delete containers.

Fixes #4882

Signed-off-by: Chen Yiyang <cyyzero@qq.com>

### CODE REVIEW

**Comment by cyyzero** [comment]
\test

File: integration/containerd/runk/runk-tests.sh [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+setup() {
+    echo "restart containerd service"
+    sudo systemctl restart containerd
+    echo "pull container image"
+    sudo ctr images pull ${TEST_IMAGE}
+}
+
+test_runk() {
+    echo "start container with runk"
+    sudo ctr run --pid-file ${CONTAINER_ID}.pid --rm -d  --runc-binary ${RUNK_BIN_PATH} ${TEST_IMAGE} ${CONTAINER_ID}
```
Comment (liubin) [inline integration/containerd/runk/runk-tests.sh:37]: It's better to configure containerd to use runk like https://github.com/kata-containers/kata-containers/blob/main/src/tools/runk/README.md#configure-containerd-to-use-runk and run
`ctr run --runtime runk ...`

File: integration/containerd/runk/runk-tests.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    read CID PID STATUS <<< $(sudo ctr t ls | grep ${CONTAINER_ID})
+    [ ${PID} == $(cat ${CONTAINER_ID}.pid) ] || die "pid is not consistent"
+    [ ${STATUS} == "RUNNING" ] || die "contianer status is not RUNNING"
+
+    echo "exec process in a container"
+    sudo ctr t exec -d --exec-id id1  ${CONTAINER_ID} sh -c "while true; do echo hello; sleep 1; done"
```
Comment (liubin) [inline integration/containerd/runk/runk-tests.sh:43]: It's better to run exec and get some results. For example:
```bash
ctr exec touch /tmp/foo
ctr exec ls /tmp/foo
```
create one file in the first exec and check it existence in the second exec.

File: integration/containerd/runk/runk-tests.sh (hunk 1 above)
Comment (cyyzero) [reply integration/containerd/runk/runk-tests.sh:37]: It seems that `ctr` and `nerdctl` ignores [plugins."io.containerd.grpc.v1.cri"] configuration because they do not use the CRI API. https://github.com/containerd/containerd/discussions/6162#discussioncomment-2032800

**Comment by cyyzero** [comment]
@liubin I have rebased this branch. Please approve and run tests again.

File: integration/containerd/runk/runk-tests.sh [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    echo "Install runk"
+    pushd ${RUNK_SRC_PATH}
+    make
+    sudo make install
+    popd
+}
+
+test_runk() {
+    echo "start container with runk"
+    sudo ctr run --pid-file ${PID_FILE} --rm -d  --runc-binary ${RUNK_BIN_PATH} ${TEST_IMAGE} ${CONTAINER_ID}
```
Comment (ManaSugi) [inline integration/containerd/runk/runk-tests.sh:52]: Nit: please remove an extra space between `-d` and `--runc-binary`.

**Comment by cyyzero** [comment]
Maybe it is caused by race condition,  since handling signal is asynced. @ManaSugi 
But how to avoid it? Sleep for a while and then check the status?

**Comment by ManaSugi** [comment]
@cyyzero Yes, it seems just race condition, please add sleep for a while (maybe about 5 seconds)  with the comment and then check the status.

**Comment by cyyzero** [comment]
@liubin Please approve and run tests

**Comment by cyyzero** [comment]
`jenkins-ci-ARM-ubuntu-18-04` and `jenkins-metrics-ubuntu-20-04` failed before entering runk tests. Please rerun them. @ManaSugi @liubin

**Comment by cyyzero** [comment]
@liubin @ManaSugi  It is able to pass all the tests now.  Do you have any other suggestions for runk tests?

**Review by ManaSugi — APPROVED** [review]

File: integration/containerd/runk/runk-tests.sh [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    [ ${STATUS} == "RUNNING" ] || die "contianer status is not RUNNING"
+
+    echo "exec process in a container"
+    sudo ctr t exec --exec-id id1 ${CONTAINER_ID} sh -c "echo hello > /tmp/foo"
+    [ "hello" == "$(sudo ctr t exec --exec-id id1 ${CONTAINER_ID} cat /tmp/foo)" ] || die "exec process failed"
+
+    echo "kill and delete container"
+    sudo ctr t kill --signal SIGKILL --all ${CONTAINER_ID}
+    # sleep for a while to make sure the task is killed
+    sleep 5s
```
Comment (liubin) [inline integration/containerd/runk/runk-tests.sh:57]: `ctr` has a global option `--timeout`, that could prevent time wasting by `sleep 5s`

File: integration/containerd/runk/runk-tests.sh [hunk 5]
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+#
+# Copyright (c) 2021-2022 Kata Contributors
```
Comment (liubin) [inline integration/containerd/runk/runk-tests.sh:3]: `2021-2022` -> `2022`

File: integration/containerd/runk/runk-tests.sh (hunk 4 above)
Comment (cyyzero) [reply integration/containerd/runk/runk-tests.sh:57]: `sudo ctr t kill --signal SIGKILL --all ${CONTAINER_ID}` to `sudo ctr --timeout 5s t kill --signal SIGKILL --all ${CONTAINER_ID}`?

Comment (GabyCT) [reply integration/containerd/runk/runk-tests.sh:57]: any reason of not using https://github.com/kata-containers/tests/blob/main/lib/common.bash#L221 @cyyzero ?

Comment (cyyzero) [reply integration/containerd/runk/runk-tests.sh:57]: @GabyCT Thanks for reminding me of this function! I will use it.

File: integration/containerd/runk/runk-tests.sh [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    echo "Install runk"
+    pushd ${RUNK_SRC_PATH}
+    make
+    sudo make install
+    popd
+}
+
+test_runk() {
+    echo "start container with runk"
+    sudo ctr run --pid-file ${PID_FILE} --rm -d --runc-binary ${RUNK_BIN_PATH} ${TEST_IMAGE} ${CONTAINER_ID}
```
Comment (GabyCT) [inline integration/containerd/runk/runk-tests.sh:45]: any reason why you are putting runc (--runc-binary) and not kata? or why not io.containerd.runc.v2?

File: integration/containerd/runk/runk-tests.sh [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+CONTAINER_ID="id1"
+PID_FILE="${CONTAINER_ID}.pid"
+
+setup() {
+    # can't find cargo when make runk below, so we install rust to make runk build correctly
+    "${dir_path}/../../../.ci/install_rust.sh" && source "$HOME/.cargo/env"
+    echo "restart containerd service"
+    sudo systemctl restart containerd
+    echo "pull container image"
+    sudo ctr images pull ${TEST_IMAGE}
```
Comment (GabyCT) [inline integration/containerd/runk/runk-tests.sh:32]: you should also use https://github.com/kata-containers/tests/blob/b77a49f6bcc4fab38a513fcf7cf9ee1e13c59d40/metrics/lib/common.bash#L67

File: integration/containerd/runk/runk-tests.sh (hunk 6 above)
Comment (cyyzero) [reply integration/containerd/runk/runk-tests.sh:45]: Just like runc, [runk](https://github.com/kata-containers/kata-containers/tree/main/src/tools/runk) is a standard OCI container runtime. It is in development and lacks testing. So I let containerd use runk to verify its functionality.

File: integration/containerd/runk/runk-tests.sh (hunk 7 above)
Comment (cyyzero) [reply integration/containerd/runk/runk-tests.sh:32]: Ok, I will use it .  But why is it under matrics directory? Maybe moving to global lib is better.

File: integration/containerd/runk/runk-tests.sh (hunk 4 above)
Comment (cyyzero) [reply integration/containerd/runk/runk-tests.sh:57]: Only SIGKILL or SIGSTOP can ensure stop the init process in a pid namespace.

According to [man 7 pid_namespaces](https://man7.org/linux/man-pages/man7/pid_namespaces.7.html), only SIGKILL or SIGSTOP can ensure stop the init process in a pid namespace.:

>        Likewise, a process in an ancestor namespace can—subject to the
>        usual permission checks described in [kill(2)](https://man7.org/linux/man-pages/man2/kill.2.html)—send signals to the
>        "init" process of a child PID namespace only if the "init"
>        process has established a handler for that signal.  (Within the
>        handler, the siginfo_t si_pid field described in [sigaction(2)](https://man7.org/linux/man-pages/man2/sigaction.2.html)
>        will be zero.)  SIGKILL or SIGSTOP are treated exceptionally:
>        these signals are forcibly delivered when sent from an ancestor
>        PID namespace.  Neither of these signals can be caught by the
>        "init" process, and so will result in the usual actions
>        associated with those signals (respectively, terminating and
>        stopping the process).

`Ctr kill` send SIGTERM by default. There is a high probability that the container process( also as the init process in PID namespace) does not establish a handler for SIGTERM. So if `ctr t kill` command (which is a process in an ancestor PID namespace) send SIGTERM, container process will not receive.

In conclusion, here should use `ctr t kill --signal SIGKILL --all`. Is it ok to update clean_env and add these options for `ctr t kill`?  I'm not sure if this will affect kata's tests.

**Comment by cyyzero** [comment]
@GabyCT Branch updated. Here must sends SIGKILL, so I doesn't use `clean_env_ctr`, but instead use `waitForProcess` directly to poll and wait for the container process receiving signal and exiting. 

cc @liubin @ManaSugi

**Comment by cyyzero** [comment]
Some tests failed. The expression in `$cmd` to the right of the equal sign is executed and expanded in advance.  `$cmd` should be a raw string and I fixed it. Please run the test again.

**Comment by cyyzero** [comment]
@GabyCT It can pass all tests now. Please review again.

**Comment by cyyzero** [comment]
Rebase and add tests for `ps` command. @liubin

**Review by liubin — APPROVED** [review]

**Comment by cyyzero** [comment]
Test fails on ubuntu 22.04 because `status::get_all_pid` only works on cgroup v1 now.  Temporarily skip tests on cgroup v2 einvitonment.  Please run test again.  @ManaSugi @liubin

**Comment by liubin** [comment]
> Test fails on ubuntu 22.04 because `status::get_all_pid` only works on cgroup v1 now. Temporarily skip tests on cgroup v2 einvitonment. Please run test again. @ManaSugi @liubin

Thank you, @cyyzero.

I created an issue to track this, please see https://github.com/kata-containers/kata-containers/issues/4821.
